src/jdk.jlink/share/classes/jdk/tools/jlink/internal/runtimelink/JimageDiffGenerator.java

```diff
@@ -126,39 +126,27 @@ public List<ResourceDiff> generateDiff(ImageResource base, ImageResource image)
     private boolean compareStreams(InputStream is1, InputStream is2) {
         byte[] buf1 = new byte[1024];
         byte[] buf2 = new byte[1024];
-        int bytesRead1, bytesRead2 = 0;
-        try {
-            try (is1; is2) {
-                while ((bytesRead1 = is1.read(buf1)) != -1 &&
-                       (bytesRead2 = is2.read(buf2)) != -1) {
-                    if (bytesRead1 != bytesRead2) {
-                        return false;
-                    }
-                    if (bytesRead1 == buf1.length) {
-                        if (!Arrays.equals(buf1, buf2)) {
-                            return false;
-                        }
-                    } else {
-                        for (int i = 0; i < bytesRead1; i++) {
-                            if (buf1[i] != buf2[i]) {
-                                return false;
-                            }
-                        }
-                    }
+        int bytesRead1, bytesRead2;
+        try (is1; is2) {
+            while ((bytesRead1 = is1.readNBytes(buf1, 0, buf1.length)) != 0) {
```

Comment:
If `in1` is exactly 1024 bytes, and `in2` is those same bytes plus one extra byte (1025 bytes), this method will return a false positive.

Probably just want `while (true)` here, then read both `in1` and `in2`, then check results.

Reply:
Good catch! Added a test for this case and fixed the code.